### PR TITLE
Commonize Scrollbar in skikoMain from desktopMain

### PR DESCRIPTION
**Pull request by eymar**
nativeMain and jsWasmMain had copy/pasted implementations.
I moved the desktop implementations into skikoMain.

The public API is preserved.

Test: tested using existing desktopTest in foundation:foundation.

At the moment, the ScrollBarTest.kt is in desktopTest only.
It's less easy to commonize because of  `@RunWith(Theories::class)`

Should I check anything else for such a change?

### CODE REVIEW

File: compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt
Comment (igordmn) [inline compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt]: How does it work on JS where we don't have ability to run something blocking?

What happens, if ScrollableState has suspensions points in `suspend fun scroll` function?

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/v2/Scrollbar.skiko.kt
Comment (igordmn) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/v2/Scrollbar.skiko.kt:28]: We should add

```
@file:JvmName("Scrollbar_desktopKt")
```

Renaming the file is not binary compatible.

File: compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt
Comment (eymar) [reply compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt]: You're right. The comment lies. 

> // This launch should return only after block is finished

To test this use case I run our example1 on desktop with:

```kotlin
    private var rawPosition: Double
        get() = scrollScale * adapter.scrollOffset
        set(value) {
            GlobalScope.launch(start = CoroutineStart.UNDISPATCHED) {
                adapter.scrollTo(value / scrollScale)
            }
//            runBlockingCoroutine {
//                adapter.scrollTo(value / scrollScale)
//            }
        }
```
When I run it on the example it looks fine. The desktop tests also pass. 

But it works now only thanks to implementation details, which can change in the future. Adding some `yield()` in the implementation breaks the tests.


> How does it work on JS where we don't have ability to run something blocking?

The current implementation of scroll finishes completely without suspensions. But runBlockingCoroutine in k/js is not blocking if something suspends.
____

Ideally, we would make runBlocking unnecessary here to have the same implementation everywhere. I'll try to think if it's possible to tweak the implementation. @m-sasha Sasha, what do you think?

Otherwise, the web implementation will be different.

Comment (igordmn) [reply compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt]: > The current implementation of scroll finishes completely without suspensions.

What if we have a custom user implementation of something scrollable, with suspension points? I didn't ivestigated it myself, when we added `runBlocking`, but that was a mistake :). Maybe we also have issues with `runBlocking` in this case.

We can investigate the `runBlocking` issue later, but replacing it with an async call probably has more consequences. Can we theoretically have a frame lag or state inconsistencies (the code waits for a state change, that didn't happen because now we are asynchronous)?

Comment (eymar) [reply compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt]: I have a draft of asynchronous implementation. I'll probably share it soon as a separate PR. 

In the worst case scenario of this Scrollbar usecase, when a user drags the scrollbar there will be a lag - it won't follow the mouse position.

In both cases either synchronous or asynchronous, if someone creates a custom implementation of something scrollable and their scroll implementation would suspend for some time/delay then the lag can be noticeable.

Comment (igordmn) [reply compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt]: > would suspend for some time/delay then the lag can be noticeable.

I haven't dived into suspention scrollables, but I heard that it uses different coroutine scope, that should perform scroll synchronously in one frame, just like in `sequence { }` + `yield`. But that just my speculation, it needs to be verified.

Comment (eymar) [reply compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt]: The draft of async soulution is here: https://github.com/JetBrains/compose-multiplatform-core/pull/575
___

> but I heard that it uses different coroutine scope, that should perform scroll synchronously in one frame

In android tests and samples they explcitily use rememberCoroutineScope() to launch scrollTo. So there's no implicit/hidden special coroutine scope for scroll. Maybe I misundertood your point.
___

In k/js we can't block. But we can attempt to finish the scroll by the end of the frame (using Microtask dispatcher), unless a custom scroll implementation dispatches new coroutines using a different Dispatcher or delays - so coroutines can be dispatched in the future frames.

So it would really depend on the scroll implementation. With current implementations in compose it would work. 

So maybe for now it's better to keep using runBlocking in jvm and native (if async solution proves to be not good enough), but use Microtask dispatcher for k/js and k/wasm

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/v2/Scrollbar.skiko.kt
Comment (MatkovIvan) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/v2/Scrollbar.skiko.kt:28]: Can we automate this check with something like https://github.com/Kotlin/binary-compatibility-validator ?

Comment (igordmn) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/v2/Scrollbar.skiko.kt:28]: Sure, the plan was eventually to add it on our CI. Until we add it, we have to track it manually

File: compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt
Comment (igordmn) [reply compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt]: > So there's no implicit/hidden special coroutine scope for scroll. Maybe I misundertood your point.

This was a guess, I don't have exact information about how it works, and what is the purpose of suspend scrollable.

Will try to dive into it more today/tomorrow.

Comment (eymar) [reply compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt]: For this PR, what do you think about:

> So maybe for now it's better to keep using runBlocking in jvm and native , but use Microtask dispatcher for k/js and k/wasm. 

So the scroll will likely finish by the next frame (not in code blocking manner though).

Comment (igordmn) [reply compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt]: After looking at the code, Compose uses `LaunchedEffect` or `rememberCoroutineScope` to scroll Scrollable via usual touch gestures.

Also, taking into account the current ComposeScene mechanic, there won't be an input lag because of `ComposeScene.postponeInvalidate`. Scroll will be scheduled first, and the next frame only after it.

Let's change it to `launch` everywhere, not only in JS, because with `runBlocking` we could encounter a deadlock (`main thread -> sendPointerEvent -> runBlocking(MainDispatcher) -> custom scrollTo -> delay(10) -> wait main thread`)

Also, it is better to make code more cleaner. Instead of mutable properties rawPosition/position we should call a function in `onDragDelta`:
```
unscrolledDragDistance += dragDelta - sliderDelta
scope.launch {
    scrollToPosition(position + sliderDelta)
}
```

cc @m-sasha, will it be okay, if we will do these changes?

Comment (igordmn) [reply compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt]: P.S. In the future, maybe we should use `FlushDispatcher` for `LaunchedEffect`'s and call `flush` after each `sendPointerEvent`, to make Composables with effects predictable. But that is another thing (there are cons/pros)

Comment (m-sasha) [reply compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt]: @igordmn It doesn't sound like a good idea, although I don't remember all the details. If the value isn't updated before the next drag event is received, it will probably cause issues.

I would like to understand why scrolling is a suspend function in the first place; it doesn't make much sense to me. Reading the docs, it seems the reason is for the scroll priority mechanism (see `DefaultScrollableState`). Maybe we should consult with our friends from Google.

Comment (igordmn) [reply compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt]: In this case we can proceed in 2 ways:

1. don't merge this PR, understand the issue more

2. use `runBlocking` in JVM/native and `launch` in web. Make Scrollbar as expect/actual (not `runblocking` itself) for marking it as Experimental on web, and create an issue to understand it later

Comment (igordmn) [reply compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt]: We can use `dispatchRawDelta` as we do with mouse wheel, but it won't work with nested scrolling. Maybe it nested scrolling issue though.

Comment (m-sasha) [reply compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt]: I don't think scrollbars need nested scrolling. In fact, I think it would be a bug (dragging an inner scrollbar shouldn't scroll the outer scrollable). But I don't think that solves the problem. We still need snapToIndex, plus it's not clear what the situation is with scroll priority/cancellation.

By the way, it seems mouse-wheel actually does need nested scrolling to work, no?

Comment (m-sasha) [reply compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt]: I think that unless we want to spend the time to understand the suspend scrolling better, the way to go is to do an expect/actual and launch it in a scope on JS.

Comment (eymar) [reply compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt]: For now, I renamed the function to `runBlockingIfPossible` to make it more clear that it can be not blocking.



> If the value isn't updated before the next drag event is received, it will probably cause issues.

@m-sasha Sasha, there is draft which attempted to solve this:
https://github.com/JetBrains/compose-multiplatform-core/pull/575 - the values should get updated before next event is processed. WDYT?

Comment (igordmn) [reply compose/foundation/foundation/src/jsWasmMain/kotlin/androidx/compose/foundation/v2/Actuals.jsWasm.kt]: > I don't think scrollbars need nested scrolling

> it seems mouse-wheel actually does need nested scrolling to work, no?

I meant the case, when we use [nestedScrollConnection](https://medium.com/@debdut.saha.1/top-app-bar-animation-using-nestedscrollconnection-like-facebook-jetpack-compose-b446c109ee52)

The case, when we just have one more scroll inside another scroll is indeed isn't needed to be handled for scrollbars. This case also already works correctly with mouse (we scroll the nested scroll)

**Comment by eymar** [comment]
@igordmn 
What do you think if we proceed with current implementation? It's synchronous for now and guaranteed to be blocking everywhere except k/js. So it doesn't change behaviour for CfD. 

We can figure out other questions llike whether blocking is appropriate or viability of async implementation out of this PR scope if needed.

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/Scrollbar.skiko.kt
Comment (MatkovIvan) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/Scrollbar.skiko.kt:1]: Please note that such rename breaks binary compatibility for all file scope public functions

Comment (igordmn) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/Scrollbar.skiko.kt:1]: `@file:JvmName("Scrollbar_desktopKt")` will fix that

**Comment by igordmn** [comment]
> We can figure out other questions llike whether blocking is appropriate or viability of async implementation out of this PR scope if needed.

I think, it is good to merge (after fixing `@JvmName`). And let's create an issue for further investigation.

**Comment by eymar** [comment]
> > We can figure out other questions llike whether blocking is appropriate or viability of async implementation out of this PR scope if needed.
> 
> I think, it is good to merge (after fixing `@JvmName`). And let's create an issue for further investigation.

Added JvmName.

`+` created a ticket to investigate - COMPOSE-150

**Review by igordmn — APPROVED** [review]

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/TextFieldScrollState.skiko.kt
Comment (MatkovIvan) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/TextFieldScrollState.skiko.kt:1]: Same issue here

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/Scrollbar.skiko.kt
Comment (MatkovIvan) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/Scrollbar.skiko.kt:1]: Not fixed in this file

**Comment by eymar** [comment]
UPD: the missing `@file:JvmName` annotations added here https://github.com/JetBrains/compose-multiplatform-core/pull/635

**Pull request merged**
